### ISSUE TITLE
Various nested optimized function invariants/errors

### ISSUE THREAD

**Issue by trueadm** [facebook/prepack#2252]
## Failure 1

Given this test case:

```js
function fn2(props) {
  var _ref11;
  var commentsConnection =
    (_ref11 = props) != null
      ? (_ref11 = _ref11.feedback) != null
        ? _ref11.display_comments
        : _ref11
      : _ref11;

  var func = props.func;

  var nested = function() {
    return func(commentsConnection);
  };
  global.__optimize && __optimize(nested);
  return props.items.map(nested);
}

function fn(props) {
  var items = Array.from(props.items);

  var func = function(commentsConnection) {
    return commentsConnection;
  };

  return fn2({
    items: items,
    func: func,
    feedback: props.feedback,
  });
}

inspect = function() {
  return JSON.stringify(fn({ items: [0, 1, 2], feedback: { display_comments: [1, 2, 3] } }))
}

global.__optimize && __optimize(fn);
```

We get the invariant:

```
Invariant Violation: 
debug-fb-www.js:206
This is likely a bug in Prepack, not your code. Feel free to open an issue on GitHub.
    at invariant (/Users/ADM/Projects/prepack/src/invariant.js:18:15)
    at Emitter.finalize (/Users/ADM/Projects/prepack/src/serializer/Emitter.js:225:15)
````

## Failure 2

A slight deviant from the last case:

```js
function fn2(props) {
  var _ref11;
  var commentsConnection =
    (_ref11 = props) != null
      ? (_ref11 = _ref11.feedback) != null
        ? _ref11.display_comments
        : _ref11
      : _ref11;

  var nested = function() {
    return props.func(commentsConnection);
  };
  global.__optimize && __optimize(nested);
  return props.items.map(nested);
}

function fn(props) {
  var items = Array.from(props.items);

  var func = function(commentsConnection) {
    return commentsConnection;
  };

  return fn2({
    items: items,
    func: func,
    feedback: props.feedback,
  });
}

inspect = function() {
  return JSON.stringify(fn({ items: [0, 1, 2], feedback: { display_comments: [1, 2, 3] } }))
}

global.__optimize && __optimize(fn);
```

The only difference is that we reference the function `props.func` argument directly rather than make a binding for it in the optimized function. Error is:

```
rror: PP1003: Property access conflicts with write in optimized function (unknown function) at 11:11 to 11:16
debug-fb-www.js:44
debug-fb-www.js:201
=== Compilation Failed ===
Error: A fatal error occurred while prepacking.
debug-fb-www.js:206
    at Functions.checkThatFunctionsAreIndependent (/Users/ADM/Projects/prepack/src/serializer/functions.js:315:13)
    at statistics.checkThatFunctionsAreIndependent.measure (/Users/ADM/Projects/prepack/src/serializer/serializer.js:142:24)
```

## Failure 3

Another failing test case:

```js
function fn2(props) {
  var _ref11;
  var commentsConnection =
    (_ref11 = props) != null
      ? (_ref11 = _ref11.feedback) != null
        ? _ref11.display_comments
        : _ref11
      : _ref11;

  var nested = function() {
    return commentsConnection;
  };
  global.__optimize && __optimize(nested);
  return props.items.map(nested);
}

function fn(props) {
  return fn2({
    items: props.items,
    feedback: props.feedback,
  });
}

inspect = function() {
  return JSON.stringify(fn({ items: [0, 1, 2], feedback: { display_comments: [1, 2, 3] } }))
}

global.__optimize && __optimize(fn);
```

This time the error is:

```
Invariant Violation: 
debug-fb-www.js:206
This is likely a bug in Prepack, not your code. Feel free to open an issue on GitHub.
    at invariant (/Users/ADM/Projects/prepack/src/invariant.js:18:15)
    at Referentializer._getReferentializationScope (/Users/ADM/Projects/prepack/src/serializer/Referentializer.js:168:15)

```

**Comment by cblappert** [comment]
For Failure 2, it seems like the message occurs because we don't know if `props` is modified by `func` beause props is passed as the `this` argument to `func`. We should probably adjust the conflict-detection logic for nested optimized functions.

I'll look into the two invariant violations.

**Comment by cblappert** [comment]
Failure 3 is because we don't visit locals of optimized functions, we assume that anything that needs the binding will visit the binding. Leaked bindings need the binding to be visited, but only visit the updated value of the binding and not the binding itself. The easiest way to address this category of issues (because there are a lot of things wrong in this area) would be a fairly large refactor that Nikolai has been wanting to do for a while.

**Comment by cblappert** [comment]
Looked into failure 2, put up #2305 for improving the error message. Looks like the creation of `func` in `fn` conflicts with the property access in `nested`.

Do we want to stop reporting or stop failing on nested optimized functions that access values created in their parent optimized functions?

**Comment by trueadm** [comment]
Did you have a response for failure 1? Maybe I missed it somewhere/somehow?

**Comment by cblappert** [comment]
Looking into it now.

**Comment by cblappert** [comment]
Fix up for failure 1 here: #2339

**Comment by cblappert** [comment]
Failure 3 fixed by #2344. Closing issue since all 3 cases have been addressed.
